### ISSUE TITLE
Automatic variance compatibility when safe

### ISSUE THREAD

**Issue by KotlinIsland** [KotlinIsland/basedmypy#410]
```py
def foo() -> list[object]:
  a = [1]
  return a
```

This code is perfectly safe and should not raise an error as it is effectively immutable as a `list[int]`.

This will involve tracking if the underlying object has any references stored as a `list[object]`
